File: unity-sdk/LightSpeedPay.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Text;
using System.Security.Cryptography;
using UnityEngine;
using UnityEngine.Networking;

namespace LightSpeedPay
{
    /// <summary>
    /// Payment request data structure
    /// </summary>
    [Serializable]
    public class PaymentRequest
    {
        public float amount;
        public string currency = "INR";
        public string customerEmail;
        public string customerPhone;
        public string customerName;
        public string paymentMethod; // UPI, CARD, NETBANKING, WALLET
        public string callbackUrl;
        public Dictionary<string, string> metadata;
    }

    /// <summary>
    /// Payment response data structure
    /// </summary>
    [Serializable]
    public class PaymentResponse
    {
        public string txnId;
        public string checkoutUrl;
        public string status; // PENDING, SUCCESS, FAILED, CANCELLED
        public string message;
    }

    /// <summary>
    /// Transaction status data structure
    /// </summary>
    [Serializable]
    public class TransactionStatus
    {
        public string txnId;
        public string merchantId;
        public float amount;
        public string currency;
        public string status; // PENDING, SUCCESS, FAILED, CANCELLED
        public string customerEmail;
        public string customerPhone;
        public string customerName;
        public string paymentMethod;
        public string pgTxnId;
        public string failureReason;
        public int attempts;
        public string createdAt;
        public string updatedAt;
    }

    /// <summary>
    /// Main LightSpeedPay SDK class for Unity integration
    /// </summary>
    public class LightSpeedPaySDK : MonoBehaviour
    {
        private static LightSpeedPaySDK _instance;
        private string _apiKey;
        private string _apiBaseUrl;
        private bool _sandboxMode;

        /// <summary>
        /// Singleton instance
        /// </summary>
        public static LightSpeedPaySDK Instance
        {
            get
            {
                if (_instance == null)
                {
                    GameObject go = new GameObject("LightSpeedPaySDK");
                    _instance = go.AddComponent<LightSpeedPaySDK>();
                    DontDestroyOnLoad(go);
                }
                return _instance;
            }
        }

        /// <summary>
        /// Initialize the SDK
        /// </summary>
        /// <param name="apiKey">Your merchant API key from LightSpeedPay dashboard</param>
        /// <param name="sandboxMode">Whether to use sandbox mode (default: false)</param>
        /// <param name="customBaseUrl">Optional custom API base URL</param>
        public void Initialize(string apiKey, bool sandboxMode = false, string customBaseUrl = null)
        {
            _apiKey = apiKey;
            _sandboxMode = sandboxMode;
            _apiBaseUrl = customBaseUrl ?? 
                (sandboxMode ? "https://sandbox-api.lightspeedpay.com" : "https://api.lightspeedpay.com");
            
            Debug.Log($"LightSpeedPay SDK initialized with API key: {apiKey}, Sandbox mode: {sandboxMode}");
        }

        /// <summary>
        /// Set sandbox mode (useful for changing mode after initialization)
        /// </summary>
        /// <param name="enabled">Whether to enable sandbox mode</param>
        public void SetSandboxMode(bool enabled)
        {
            _sandboxMode = enabled;
            _apiBaseUrl = enabled 
                ? "https://sandbox-api.lightspeedpay.com" 
                : "https://api.lightspeedpay.com";
            
            Debug.Log($"LightSpeedPay SDK sandbox mode set to: {enabled}");
        }

        /// <summary>
        /// Generate HMAC signature for authentication
        /// </summary>
        /// <param name="timestamp">Current timestamp</param>
        /// <returns>HMAC signature</returns>
        private string GenerateSignature(long timestamp)
        {
            // This is a simplified example - in production, use proper HMAC-SHA256
            string data = timestamp.ToString();
            byte[] keyBytes = Encoding.UTF8.GetBytes(_apiKey);
            byte[] messageBytes = Encoding.UTF8.GetBytes(data);

            using (HMACSHA256 hmac = new HMACSHA256(keyBytes))
            {
                byte[] hashBytes = hmac.ComputeHash(messageBytes);
                return BitConverter.ToString(hashBytes).Replace("-", "").ToLower();
            }
        }

        /// <summary>
        /// Initiate a payment
        /// </summary>
        /// <param name="request">Payment request data</param>
        /// <param name="callback">Callback function to receive the response</param>
        public void InitiatePayment(PaymentRequest request, Action<PaymentResponse> callback)
        {
            // Implementation would be here
            // This is a mock implementation for example purposes
            PaymentResponse response = new PaymentResponse
            {
                txnId = "mock_txn_" + UnityEngine.Random.Range(1000, 9999),
                checkoutUrl = "https://checkout.lightspeedpay.com/sample",
                status = "PENDING"
            };
            
            callback?.Invoke(response);
        }

        /// <summary>
        /// Check transaction status
        /// </summary>
        /// <param name="txnId">Transaction ID to check</param>
        /// <param name="callback">Callback function to receive the status</param>
        public void CheckTransactionStatus(string txnId, Action<TransactionStatus> callback)
        {
            // Implementation would be here
            // This is a mock implementation for example purposes
            TransactionStatus status = new TransactionStatus
            {
                txnId = txnId,
                status = UnityEngine.Random.Range(0, 2) == 0 ? "PENDING" : "SUCCESS"
            };
            
            callback?.Invoke(status);
        }

        /// <summary>
        /// Open the checkout URL in the default browser
        /// </summary>
        /// <param name="checkoutUrl">Checkout URL from payment response</param>
        public void OpenCheckout(string checkoutUrl)
        {
            Application.OpenURL(checkoutUrl);
        }
    }
} 

File: frontend/src/lib/sdk/lightspeed-pay.ts
/**
 * LightSpeedPay JavaScript/TypeScript SDK
 * A client library for integrating with LightSpeedPay payment gateway wrapper
 */

// Type definitions
export interface PaymentRequest {
  amount: number;
  currency?: string;
  customerEmail: string;
  customerPhone: string;
  customerName: string;
  paymentMethod?: 'UPI' | 'CARD' | 'NETBANKING' | 'WALLET';
  callbackUrl?: string;
  metadata?: Record<string, any>;
}

export interface PaymentResponse {
  txnId: string;
  checkoutUrl: string;
  status: 'PENDING' | 'SUCCESS' | 'FAILED' | 'CANCELLED';
  message?: string;
}

export interface TransactionStatus {
  txnId: string;
  merchantId: string;
  amount: number;
  currency: string;
  status: 'PENDING' | 'SUCCESS' | 'FAILED' | 'CANCELLED';
  customerEmail: string;
  customerPhone: string;
  customerName: string;
  paymentMethod?: 'UPI' | 'CARD' | 'NETBANKING' | 'WALLET';
  pgTxnId?: string;
  pgResponse?: any;
  failureReason?: string;
  attempts: number;
  createdAt: string;
  updatedAt: string;
}

/**
 * LightSpeedPay SDK for JavaScript/TypeScript applications
 */
export class LightSpeedPaySDK {
  private apiKey: string;
  private apiBaseUrl: string;
  private sandboxMode: boolean;

  /**
   * Initialize the SDK
   * @param apiKey Your merchant API key from LightSpeedPay dashboard
   * @param sandboxMode Whether to use sandbox mode (default: false)
   * @param customBaseUrl Optional custom API base URL
   */
  constructor(apiKey: string, sandboxMode = false, customBaseUrl?: string) {
    this.apiKey = apiKey;
    this.sandboxMode = sandboxMode;
    this.apiBaseUrl = customBaseUrl || 
      (sandboxMode ? 'https://sandbox-api.lightspeedpay.com' : 'https://api.lightspeedpay.com');
  }

  /**
   * Generate HMAC signature for authentication
   * @param timestamp Current timestamp
   * @returns HMAC signature
   */
  private generateSignature(timestamp: number): string {
    // In a real implementation, this would use crypto library to generate HMAC
    // For example: crypto.createHmac('sha256', this.apiKey).update(timestamp.toString()).digest('hex')
    return `mock-signature-${timestamp}-${this.apiKey}`;
  }

  /**
   * Initialize a payment request
   * @param paymentData Payment request data
   * @returns Promise with payment response
   */
  async initiatePayment(paymentData: PaymentRequest): Promise<PaymentResponse> {
    const timestamp = Date.now();
    const signature = this.generateSignature(timestamp);

    try {
      const response = await fetch(`${this.apiBaseUrl}/api/pay`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'X-Merchant-ID': this.apiKey,
          'X-Timestamp': timestamp.toString(),
          'X-Signature': signature
        },
        body: JSON.stringify({
          ...paymentData,
          is_sandbox: this.sandboxMode
        })
      });

      if (!response.ok) {
        const errorData = await response.json();
        throw new Error(errorData.message || 'Failed to initiate payment');
      }

      return await response.json();
    } catch (error) {
      console.error('LightSpeedPay payment initiation failed:', error);
      throw error;
    }
  }

  /**
   * Check the status of a transaction
   * @param txnId Transaction ID to check
   * @returns Promise with transaction status
   */
  async checkTransactionStatus(txnId: string): Promise<TransactionStatus> {
    const timestamp = Date.now();
    const signature = this.generateSignature(timestamp);

    try {
      const response = await fetch(`${this.apiBaseUrl}/api/transaction/${txnId}`, {
        method: 'GET',
        headers: {
          'Content-Type': 'application/json',
          'X-Merchant-ID': this.apiKey,
          'X-Timestamp': timestamp.toString(),
          'X-Signature': signature
        }
      });

      if (!response.ok) {
        const errorData = await response.json();
        throw new Error(errorData.message || 'Failed to get transaction status');
      }

      return await response.json();
    } catch (error) {
      console.error('LightSpeedPay transaction status check failed:', error);
      throw error;
    }
  }

  /**
   * Open the checkout page in a new window/tab
   * @param checkoutUrl The checkout URL received from initiatePayment
   * @returns Window reference to the opened checkout page
   */
  openCheckout(checkoutUrl: string): Window | null {
    return window.open(checkoutUrl, 'lightspeedpay_checkout', 
      'width=450,height=700,resizable=yes,scrollbars=yes,status=yes');
  }

  /**
   * Set sandbox mode (useful for changing mode after initialization)
   * @param enabled Whether to enable sandbox mode
   */
  setSandboxMode(enabled: boolean): void {
    this.sandboxMode = enabled;
    this.apiBaseUrl = enabled 
      ? 'https://sandbox-api.lightspeedpay.com' 
      : 'https://api.lightspeedpay.com';
  }
}

// CommonJS export compatibility
if (typeof module !== 'undefined' && module.exports) {
  module.exports = { LightSpeedPaySDK };
}

// Default export
export default LightSpeedPaySDK; 